src/hardware/mixer.cpp

```diff
@@ -263,29 +263,64 @@ int MIXER_GetSampleRate()
 	return sample_rate_hz;
 }
 
-void MIXER_LockAudioDevice()
+bool MIXER_InChipmunkMode()
 {
-	if (mixer.sdl_device > 0) {
-		SDL_LockAudioDevice(mixer.sdl_device);
+	return mixer.chipmunk_mode;
+}
+
+// The queues listed here are for audio devices that run on the main thread
+// The mixer thread can be waiting on the main thread to produce audio in these queues
+// We need to stop them before aquiring a mutex lock to avoid a deadlock
+// These are called infrequently when global mixer state is changed (mostly on device init/destroy and in the MIXER command line program)
+// Individual channels also have a mutex which can be safely aquired without stopping these queues
+void MIXER_LockMixerThread()
+{
+	if (pc_speaker) {
+		pc_speaker->output_queue.Stop();
 	}
+	if (tandy_dac) {
+		tandy_dac->output_queue.Stop();
+	}
+	if (ps1_dac) {
+		ps1_dac->output_queue.Stop();
+	}
+	if (lpt_dac) {
+		lpt_dac->output_queue.Stop();
+	}
+	soundblaster_mixer_queue.Stop();
+	reel_magic_audio.output_queue.Stop();
+	mixer.mutex.lock();
 }
 
-void MIXER_UnlockAudioDevice()
+void MIXER_UnlockMixerThread()
 {
-	if (mixer.sdl_device > 0) {
-		SDL_UnlockAudioDevice(mixer.sdl_device);
+	if (pc_speaker) {
+		pc_speaker->output_queue.Start();
+	}
+	if (tandy_dac) {
+		tandy_dac->output_queue.Start();
+	}
+	if (ps1_dac) {
+		ps1_dac->output_queue.Start();
 	}
+	if (lpt_dac) {
+		lpt_dac->output_queue.Start();
+	}
+	soundblaster_mixer_queue.Start();
+	reel_magic_audio.output_queue.Start();
+	mixer.mutex.unlock();
```

Comment:
don't you typically release a mutex in the reverse order you acquire it? So if you Stop all the queues and then Lock, don't you Unlock first and then do the Starts?

Reply:
Unlocking first could have the mixer playback a block of silence.  Mixer thread starts running through the channels, tries to dequeue but fails because it's stopped so it just plays back silence.  I think this is the correct order.